yxi/xml/xml_generator.py

```diff
@@ -0,0 +1,113 @@
+import sys
+import json
+from xml.etree.ElementTree import Element, SubElement, tostring
+from xml.dom import minidom
+from math import log2
+
+"""
+This file takes in a `.yxi` description and outputs a xml suitable for a `kernel.xml` file
+can be used to package an xclbin for the Xilinx XRT runtime.
+See https://docs.amd.com/r/en-US/ug1393-vitis-application-acceleration/RTL-Kernel-XML-File
+for the spec this is based on.
+"""
+size_key = "total_size"
+width_key = "data_width"
+
+def gen_xml(yxi):
+  mems = yxi["memories"]
+  check_mems_wellformed(mems)
+
+  root = Element("root", {"versionMajor": "1", "versionMinor": "6"})
+  kernel = SubElement(root, "kernel", {
+    "name" : yxi["toplevel"],
+    "language": "ip_c",
+    # TODO: Make sure this matches component.xml, Namely the `Toplevel` part.
+    #  See https://docs.amd.com/r/en-US/ug1393-vitis-application-acceleration/RTL-Kernel-XML-File
+    "vlnv": "capra.cs.cornell.edu:kernel:Toplevel:1.0",
+    "attributes": "",
+    "preferredWorkGroupSizeMultiple" : "0",
+    "workGroupSize" : "1",
+    "hwControlProtocol": "ap_ctrl_hs"
+  })
+
+  # Construct ports
+  ports = SubElement(kernel, "ports")
+  # The subordinates XRT - AXI controller is added outside of the programs memory interface.
+  SubElement(ports, "port", {
+    "name": "S_AXI_CONTROL",
+    "mode": "slave",
+    "range": "0x1000", # XXX: Why is this 0x1000? Taken from the Xilinx examples  Maybe could be dynamically calculated?
+    "dataWidth" : "32",
+    "portType" : "addressable",
+    "base" : "0x0"
+  })
+
+  for mem in mems:
+    SubElement(ports, "port", {
+      "name": f"m_axi_{mem['name']}",
+      "mode": "master",
+      #NOTE(nathaniel): In the Xilinx examples range is usually 0xFFFFFF... but this should be fine for us?
+      "range": f"{hex(size_in_bytes(mem))}",
+      "dataWidth": f"{mem[width_key]}",
+      "portType": "addressable",
+      "base": "0x0"
+    })
+
+  # Construct Args
+  args = SubElement(kernel, "args")
+  # XRT spec starts args addresses at 0x10
+  args_addr = 0x10
+  for i, mem in enumerate(mems):
+    SubElement(args, "arg", {
+      "name": f"{mem['name']}",
+      "addressQualifier": "1", # 1 denotes the arguments as a global memory, 
+      "id" : f"{i}",
+      "port" : f"m_axi_{mem['name']}",
+      "size" : f"0x8", # XRT expects AXI manager interfaces that are 64 bits wide 
+      "offset" : f"{hex(args_addr + (i * 8))}",
+      "type" : "int*", # TODO: Old xml generator defaults to `int*`, I imagine in practice we want to get this from our .dat
```

Comment:
I'd actually argue that hard-coding `int*` is probably the right thing… in the sense that Calyx is currently totally agnostic to the interpretation of the bits, so `float` and `int` are treated the same way. Until we break with that and want to start enriching the interface information with datatype stuff, `int*` is a suitable way to tell XRT "this is just a bag of bits."